File: Sorting_and_Searching/Collecting_Numbers.cpp
#include <bits/stdc++.h>

using namespace std;

struct comp {
    bool operator()(const pair<int, int>& a, const pair<int, int>& b) const {
        return a.first < b.first;
    }
};

int collectingNumbers(int n, vector<pair<int, int>> integers) {
    int rounds = 0;
    int i = 0;
    
    while (i < n) {
        while (integers[i + 1].second > integers[i].second) {
            i++;
        }
        i++;
        rounds++;
    }

    return rounds;
}


int main() {
    int n;
    cin >> n;

    vector<pair<int, int>> integers;
    for (int i = 1; i <= n; i++) {
        int intg;
        cin >> intg;
        integers.emplace_back(make_pair(intg, i));
    }

    sort(integers.begin(), integers.end(), comp());
    cout << collectingNumbers(n, integers);
    return 0;
}

File: Sorting_and_Searching/Collecting_Numbers.py
from dataclasses import dataclass
@dataclass
class Pair:
    first : int
    second : int
    
def collectingNumbers(numsPairs : list[Pair], n : int) :
    i : int = 0
    rounds : int = 0
    while (i < n) :
        while (i < n - 1 and numsPairs[i].second < numsPairs[i + 1].second) :
            i += 1    
        
        rounds += 1
        i += 1
    
    print(rounds)

if __name__ == '__main__' :
    n : int = int(input())
    numsInput : list[int] = list(map(int, input().split()) )
    numsPairs : list[Pair] = sorted([Pair(numsInput[i], i) for i in range(0, len(numsInput))], key=lambda p : p.first)
    collectingNumbers(numsPairs, n)
